server/clone_grpc_test.go

```diff
@@ -0,0 +1,380 @@
+package server
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"math"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/cybozu-go/moco"
+	"github.com/cybozu-go/moco-agent/metrics"
+	"github.com/cybozu-go/moco-agent/server/proto"
+	"github.com/cybozu-go/moco-agent/test_utils"
+	"github.com/cybozu-go/moco/accessor"
+	"github.com/google/go-cmp/cmp"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+func testClonegRPC() {
+	var agent *Agent
+	var registry *prometheus.Registry
+	var gsrv proto.CloneServiceServer
+
+	BeforeEach(func() {
+		// The configuration of the donor MySQL is different for each test case.
+		// So the donor is not initialized here. The initialization will do at the beginning of each test case.
+		test_utils.StopAndRemoveMySQLD(donorHost)
+		test_utils.StopAndRemoveMySQLD(replicaHost)
+
+		err := test_utils.StartMySQLD(replicaHost, replicaPort, replicaServerID)
+		Expect(err).ShouldNot(HaveOccurred())
+		err = test_utils.InitializeMySQL(replicaPort)
+		Expect(err).ShouldNot(HaveOccurred())
+		err = test_utils.SetValidDonorList(replicaPort, donorHost, donorPort)
+		Expect(err).ShouldNot(HaveOccurred())
+		err = test_utils.ResetMaster(replicaPort)
+		Expect(err).ShouldNot(HaveOccurred())
+
+		registry = prometheus.NewRegistry()
+		metrics.RegisterMetrics(registry)
+
+		agent = New(test_utils.Host, clusterName, token, test_utils.MiscUserPassword, test_utils.CloneDonorUserPassword, replicationSourceSecretPath, "", replicaPort,
+			&accessor.MySQLAccessorConfig{
+				ConnMaxLifeTime:   30 * time.Minute,
+				ConnectionTimeout: 3 * time.Second,
+				ReadTimeout:       30 * time.Second,
+			},
+		)
+
+		gsrv = NewCloneService(agent)
+	})
+
+	AfterEach(func() {
+		test_utils.StopAndRemoveMySQLD(donorHost)
+		test_utils.StopAndRemoveMySQLD(replicaHost)
+	})
+
+	It("should return error with bad requests", func() {
+		initializeDonorMySQL(false)
+
+		testcases := []struct {
+			title string
+			req   *proto.CloneRequest
+		}{
+			{
+				title: "passing invalid token",
+				req: &proto.CloneRequest{
+					Token:     "invalid-token",
+					DonorHost: donorHost,
+					DonorPort: donorPort,
+				},
+			},
+			{
+				title: "passing empty token",
+				req: &proto.CloneRequest{
+					DonorHost: donorHost,
+					DonorPort: donorPort,
+				},
+			},
+			{
+				title: "passing empty donorHostName",
+				req: &proto.CloneRequest{
+					Token:     token,
+					DonorPort: donorPort,
+				},
+			},
+			{
+				title: "passing empty donorPort",
+				req: &proto.CloneRequest{
+					Token:     token,
+					DonorHost: donorHost,
+				},
+			},
+		}
+
+		for _, tt := range testcases {
+			By(tt.title)
+			_, err := gsrv.Clone(context.Background(), tt.req)
+			Expect(err).Should(HaveOccurred())
+		}
+
+		By("checking metrics")
+		// In these test cases, the clone will not start actually. So the metrics will not change.
+		_, err := getMetric(registry, metricsPrefix+"clone_count")
```

Comment:
Why don't you check the value of the metrics?

Reply:
I added some description to `getMetrics()`.